api/app/Listeners/ComputeFinalDecisionAndCurrentStep.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace App\Listeners;
+
+use App\Enums\AssessmentDecision;
+use App\Events\AssessmentResultSaved;
+
+class ComputeFinalDecisionAndCurrentStep
+{
+    /**
+     * Create the event listener.
+     */
+    public function __construct()
+    {
+        //
+    }
+
+    /**
+     * Handle the event.
+     */
+    public function handle(AssessmentResultSaved $event): void
+    {
+        $result = $event->result;
+        $result->load([
+            'poolCandidate',
+            'poolCandidate.pool.assessmentSteps',
+        ]);
+
+        $candidate = $result->poolCandidate;
+
+        $decisionsPerStep = $candidate->computeDecisionsPerAssessmentStep();
+
+        $currentStep = 1;
+        $orderedSteps = $candidate->pool->assessmentSteps()->orderBy('sort_order', 'ASC')->get();
+        foreach ($orderedSteps as $step) {
```

Comment:
Didn't we add some fancy logic for how HOLDs interacted with current step. I think we landed on "if the final step is HOLD, treat it like TO ASSESS". But I'll confirm...

Reply:
Okay, so here's the logic we currently have on the frontend. As I thought, we treat a HOLD on the final step as if it were TO ASSESS.

Oh, and we use zero indexing. 🤦‍♂️ Not sure what to make of that.

```
/**
 * Returns the "current step" a candidate should appear at in the assessment tracker.
 * A candidate should appear in all columns less than or equal to its current step.
 * A value of null means that a candidate has passed all steps.
 * @param assessmentToResult
 * @param assessmentOrdering
 * @returns
 */
const determineCurrentStep = (
  assessmentToResult: Map<AssessmentStepId, NullableDecision>,
  assessmentOrdering: string[],
): number | null => {
  for (let index = 0; index < assessmentOrdering.length; index += 1) {
    const assessmentStepId = assessmentOrdering[index];
    const result = assessmentToResult.get(assessmentStepId);
    if (result === AssessmentDecision.Unsuccessful || result === NO_DECISION) {
      return index;
    }
    // A candidate can be qualified with some Hold decisions, as long as they are followed by a Successful decision.
    // That means that if the final step is Hold, we treat it more like NO_DECISION.
    if (
      index === assessmentOrdering.length - 1 &&
      result === AssessmentDecision.Hold
    ) {
      return index;
    }
  }
  return null;
};
```